### PR TITLE
Use hierarchies to build definitions

### PR DESCRIPTION
**Pull request by inglesp**
Suggest reviewing commit-by-commit.

### CODE REVIEW

File: codelists/definition.py [hunk 1]
```diff
@@ -29,62 +34,104 @@ def __hash__(self):
...
+    def __init__(self, rules):
+        self.rules = set(rules)
 
-    def unnegated_elements(self):
-        for e in self.elements:
-            if not e.negated:
-                yield e
+    def excluding_rules(self):
+        """Yield rules that exclude a code (and possible that code's descendants).
+        """
```
Comment (ghickman) [inline codelists/definition.py]: Single line docstrings have the quotes on the same line.

File: codelists/definition.py [hunk 2]
```diff
@@ -29,62 +34,104 @@ def __hash__(self):
...
+
+        for rule in self.rules:
+            if rule.code_is_excluded:
+                yield rule
+
+    def including_rules(self):
+        """Yield rules that include a code (and possible that code's descendants).
+        """
```
Comment (ghickman) [inline codelists/definition.py]: Single line docstrings have the quotes on the same line.

File: codelists/definition.py [hunk 3]
```diff
@@ -29,62 +34,104 @@ def __hash__(self):
...
+
+        for rule in self.rules:
+            if not rule.code_is_excluded:
+                yield rule
 
     def codes(self, tree):
+        """Return set of codes that is defined by the rules.
+        """
```
Comment (ghickman) [inline codelists/definition.py:32]: Single line docstrings have the quotes on the same line.

File: codelists/definition.py [hunk 4]
```diff
@@ -65,50 +62,61 @@ def from_query(cls, fragments):
...
 
-        This code will change significantly in an upcoming commit, when it will have a
-        proper docstring.
+        * If all the node's descendants are included in the set, we add a DefinitionRule
+        that applies to that code and all its descendants.
+
+        * If the proporition of a code's descendants that are missing from the
+        set of codes is less than r, we add a DefinitionRule that applies to that code
+        and all its descendants, and add extra rules that exclude the descendants that
+        are missing.
```
Comment (ghickman) [inline codelists/definition.py:82]: This is a brilliant explanation 👍

**Review by ghickman — APPROVED** [review]
👍

File: codelists/definition.py (hunk 1 above)
Comment (inglesp) [reply codelists/definition.py]: Tell that to the linter!

File: codelists/definition.py (hunk 4 above)
Comment (inglesp) [reply codelists/definition.py:82]: Thanks :smile:

**Pull request merged**